src/utils/circular_array_buffer.jl

```diff
@@ -122,40 +119,31 @@ end
 
 Base.IndexStyle(::CircularArrayBuffer) = IndexLinear()
 Base.size(cb::CircularArrayBuffer{<:Any,N}, i::Integer) where {N} =
-    i == N ? cb.length : size(cb.buffer, i)
-Base.size(cb::CircularArrayBuffer{<:Any,N}) where {N} = ntuple(M -> size(cb, M), N)
+    i == N ? cb.nframes : size(cb.buffer, i)
+Base.size(cb::CircularArrayBuffer{<:Any,N}) where {N} = ntuple(i -> size(cb, i), N)
 Base.getindex(cb::CircularArrayBuffer{T,N}, i::Int) where {T,N} =
     getindex(cb.buffer, _buffer_index(cb, i))
 Base.setindex!(cb::CircularArrayBuffer{T,N}, v, i::Int) where {T,N} =
     setindex!(cb.buffer, v, _buffer_index(cb, i))
 capacity(cb::CircularArrayBuffer{T,N}) where {T,N} = size(cb.buffer, N)
-isfull(cb::CircularArrayBuffer) = cb.length == capacity(cb)
-Base.isempty(cb::CircularArrayBuffer) = cb.length == 0
+isfull(cb::CircularArrayBuffer) = cb.nframes == capacity(cb)
+Base.isempty(cb::CircularArrayBuffer) = cb.nframes == 0
 
 @inline function _buffer_index(cb::CircularArrayBuffer, i::Int)
     ind = (cb.first - 1) * cb.step_size + i
-    if ind > length(cb.buffer)
-        ind - length(cb.buffer)
-    else
-        ind
-    end
+    mod1(ind, length(cb.buffer))
```

Comment:
Just spent a whole night debugging a huge performance degrade and finally get the answer. 😭 😭 😭

Reply:
Oh... I'm Sorry. I didn't know using `mod1` would cause big performance degradation.
What should be done if `ind` exceeds `2n`?

Reply:
`ind` is guaranteed to not exceed `2n` by bound check.

Reply:
You may add test cases if you're still unsure.

Reply:
Ok